File: apps/mobile/lib/core/models/composer_mcp.dart
import 'mcp_models.dart';

String sanitizeMcpServerName(String name) {
  final normalized = name
      .trim()
      .toLowerCase()
      .replaceAll(RegExp(r'[^a-z0-9_-]+'), '-')
      .replaceAll(RegExp(r'^-+|-+$'), '');
  return normalized.isEmpty ? 'mcp-server' : normalized;
}

List<String> listEnabledGlobalMcpServerKeys(List<McpServerConfigView> servers) {
  return servers
      .where((server) => server.enabled && server.name.trim().isNotEmpty)
      .map((server) => sanitizeMcpServerName(server.name))
      .toList(growable: false);
}

Map<String, bool> deriveMcpServersEnabled(
  List<String> availableServerKeys, {
  List<String> profileAssignedServers = const [],
  Map<String, bool>? existing,
  Map<String, bool>? remembered,
}) {
  final profileAssigned = profileAssignedServers
      .map(sanitizeMcpServerName)
      .toSet();
  final result = <String, bool>{};
  for (final key in availableServerKeys) {
    final sanitized = sanitizeMcpServerName(key);
    final existingValue = existing?[sanitized];
    if (existingValue is bool) {
      result[sanitized] = existingValue;
      continue;
    }
    final rememberedValue = remembered?[sanitized];
    if (rememberedValue is bool) {
      result[sanitized] = rememberedValue;
      continue;
    }
    result[sanitized] = profileAssigned.contains(sanitized);
  }
  return result;
}

List<String> resolveEnabledMcpServerKeys(Map<String, bool> settings) {
  return settings.entries
      .where((entry) => entry.value)
      .map((entry) => entry.key)
      .toList(growable: false);
}

int countEnabledMcpServers(Map<String, bool> settings) {
  return settings.values.where((enabled) => enabled).length;
}

Map<String, bool>? normalizeMcpServersEnabled(Map<String, dynamic>? input) {
  if (input == null || input.isEmpty) return null;
  final result = <String, bool>{};
  for (final entry in input.entries) {
    final key = entry.key.trim();
    final value = entry.value;
    if (key.isNotEmpty && value is bool) {
      result[sanitizeMcpServerName(key)] = value;
    }
  }
  return result.isEmpty ? null : result;
}

File: apps/desktop/src/shared/composer-mcp.ts
import { sanitizeMcpServerName } from "./mcp";
import type { McpServerConfigView } from "./ipc";

export type McpServersEnabledSettings = Record<string, boolean>;

export function listEnabledGlobalMcpServerKeys(servers: readonly McpServerConfigView[]): string[] {
  return servers
    .filter((server) => server.enabled && server.name.trim())
    .map((server) => sanitizeMcpServerName(server.name));
}

export function deriveMcpServersEnabled(
  availableServerKeys: readonly string[],
  options: {
    profileAssignedServers?: readonly string[];
    existing?: Partial<McpServersEnabledSettings>;
    remembered?: Partial<McpServersEnabledSettings>;
  } = {},
): McpServersEnabledSettings {
  const profileAssigned = new Set(
    (options.profileAssignedServers ?? []).map((server) => sanitizeMcpServerName(server)),
  );
  const result: McpServersEnabledSettings = {};
  for (const key of availableServerKeys) {
    const sanitized = sanitizeMcpServerName(key);
    if (typeof options.existing?.[sanitized] === "boolean") {
      result[sanitized] = options.existing[sanitized];
      continue;
    }
    if (typeof options.remembered?.[sanitized] === "boolean") {
      result[sanitized] = options.remembered[sanitized];
      continue;
    }
    result[sanitized] = profileAssigned.has(sanitized);
  }
  return result;
}

export function resolveEnabledMcpServerKeys(settings: McpServersEnabledSettings): string[] {
  return Object.entries(settings)
    .filter(([, enabled]) => enabled)
    .map(([key]) => key);
}

export function countEnabledMcpServers(settings: McpServersEnabledSettings): number {
  return Object.values(settings).filter(Boolean).length;
}

export function normalizeMcpServersEnabled(value: unknown): McpServersEnabledSettings | undefined {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return undefined;
  }
  const result: McpServersEnabledSettings = {};
  for (const [key, enabled] of Object.entries(value)) {
    if (typeof enabled === "boolean" && key.trim()) {
      result[sanitizeMcpServerName(key)] = enabled;
    }
  }
  return Object.keys(result).length > 0 ? result : undefined;
}
